redb-derive/src/lib.rs

```diff
@@ -0,0 +1,54 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, DeriveInput};
+
+#[proc_macro_derive(RedbValue, attributes(fixed_width, type_name))]
+pub fn redb_value(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = input.ident;
+
+    let mut width = quote!(None);
+    let mut type_name = quote!(stringify!(#name));
+    for attr in input.attrs {
+        if attr.path().is_ident("fixed_width") {
+            width = quote!(Some(std::mem::size_of::<Self>()));
+        }
+        if attr.path().is_ident("type_name") {
+            type_name = attr.meta.require_list().unwrap().tokens.clone()
+        }
+    }
+
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+
+    let expanded = quote! {
+        impl #impl_generics redb::RedbValue for #name #ty_generics #where_clause {
+            type SelfType<'a> = Self;
+            type AsBytes<'a> = Vec<u8> where Self: 'a;
+
+            fn fixed_width() -> Option<usize> {
+                #width
+            }
+
+            fn from_bytes<'a>(data: &'a [u8]) -> Self
+            where
+                Self: 'a,
+            {
+                redb::deserialize(data).unwrap()
+            }
+
+            fn as_bytes<'a, 'b: 'a>(value: &'a Self::SelfType<'b>) -> Self::AsBytes<'a>
+            where
+                Self: 'a,
+                Self: 'b,
+            {
+                redb::serialize(value).unwrap()
+            }
+
+            fn type_name() -> redb::TypeName {
+                redb::TypeName::new(#type_name)
```

Comment:
I've never written a proc macro before. Is it possible to change this to include the types of all the fields in the struct? For example, the way I [did it for tuples](https://github.com/cberner/redb/blob/505e21915b681ac4e409b9b8e5b280ed5657d55e/src/tuple_types.rs#L75-L90). This would resolve my concern about users changing the type of a field and then getting corrupted data instead of an error. Something like `#type_name {#field_1_type, #field_2_type, #field_3_type...}`

Reply:
Definitely possible.

BTW. This is stored only once per table, right? I guess my worries were needless if so.

Reply:
Yes, once per table. And it's only checked once when the table is opened, so shouldn't add much overhead if it's long

Reply:
In that case a must have.